oeps/oep-0032-arch-unique-identifier-for-users.rst

```diff
@@ -0,0 +1,125 @@
+
+.. _Django auth_user: https://docs.djangoproject.com/en/2.0/topics/auth/default/#user-objects
+
+
+Context
+=======
+
+In 2016, the analytics team decided to use the the *LMS user_id* as a consistent user identifier for analytics events, and this Open edX Proposal (OEP) further documents and builds upon that decision.
+
+In addition to needing a consistent user identifier for internal events, we also need one for public events, REST APIs and JavaScript APIs.
+
+
+User Identifier Requirements
+----------------------------
+
+The requirements for a User Identifier are as follows:
+
+* Uniqueness: the identifier must not represent more than one user.
+
+* The identifier must map back to other information about the user (e.g. demographics, enrollments, etc.).
+
+* The identifier for a given user should not change.
+
+* The identifier should not contain `Personally Identifiable Information (PII)`_.
+
+* The identifier should be consistent across all systems, including the LMS and Studio, mobile apps for iOS and Android, and other Open edX Services like Ecommerce.
+
+.. _Personally Identifiable Information (PII): oep-0030-arch-pii-markup-and-auditing.rst
+
+
+User Identifier Candidate Comparison
+------------------------------------
+
+The following table summarizes the comparisons of a list of candidate user identifiers.
+
+.. list-table::
+   :header-rows: 1
+   :widths: 25 75
+
+   * - Identifier
+     - Comments
+   * - username
+     - * Username is considered PII.
+   * - LMS user_id
+     - * anonymous (is not considered PII)
+
+       * Segment (used for anlytics) recommends database IDs because they never change.
+
+       * Availability across all services is almost, but not yet complete.
+
+       * Auto-incremented id allows for possible information leakage of auth_user table and ties us down to implementation of the housing database.
+   * - email address
+     - * Is PII and modifiable by the user.
+   * - anonymous user_id
+     - * anonymous (is not considered PII)
+
+       * It is currently constructed by hashing the user's *LMS user_id* with the Django server's *SECRET_KEY* value. This value would change when the *SECRET_KEY* is rotated.
+
+As noted above, the *LMS user_id* was selected for internal analytics events as the best match given the requirements. It continues to be the best match for the additional use cases of public events, REST APIs and JavaScript APIs.
+
+
+Consequences
+============
+
+A consequence of the decision to use the *LMS user_id* as the unique user id is that some additional changes are needed to make the *LMS user_id* more globally available across all services:
+
+* The `LMS OAuth Dispatch App`_ will now include the *LMS user_id* in the JWTs (JSON Web Tokens) used as part of the `Open edX Authentication`_ upon request.
+
+* Additionally, some further investment is required to make the *LMS user_id* available to all services for all users, like the Ecommerce Service.
```

Comment:
Yes. We are a lot further, but there is additional work that I think the Revenue team plans to pick up. At some point, we hope this will be resolved in all cases.

Reply:
Leaving for now.  We can update later when this is no longer needed.